### DISCUSSION TITLE
Constraint on two axes (2D physics simulation)

### DISCUSSION THREAD

**Discussion by xcasadio** [bepu/bepuphysics2#255] [Q&A]
Hello.

I could not find in the demos or in the code the possibility to put a constraint on the axes.
In some physics engine you can define a constraint on an axis. For example if you create a scrolling game on X and Y, you can't move on the Z axis.

I looked at the constraints implemented in BepuPhysics.Constraints but I didn't see anything like that.

Does it already exist or do I have to create a new constraint?

**Comment by RossNordby** [comment]
`LinearAxisServo` would work (the second connection can be a 'world' kinematic body), but a constraint may not actually be necessary here. Simply setting the position/velocity every timestep may be sufficient, depending on the requirements.

Either way, be advised that this doesn't totally delete the third degree of freedom, and sufficient force could smash two objects into each other such that the shortest path to separation is along the restricted axis. In that case, the colliders won't separate in the 2D plane. To avoid this, you can make sure the objects have sufficient depth that the restricted axis is never a good choice, or detect such collisions based on the contact normal and handle them directly.

**Comment by damian-666** [comment]
Wow this got me thinking deep, I do 2d physics game as well.  but now I'm
thinking in voxel space  like a some of the more logical quantum theorists
, this is because  ..I have to  use eulerian Fluids in discreet density
fields and rigid bodies interaction with either sdf or the box2d way and
some chimera method  to put those 2 way interactions, an adhoc invention
thet is ok.  So I'm mixing discreet eulerian desnity grids, and field
models.    They leak pessure masss, and energy thouhh. But stable ansit's
ok for a game.

But this question give me an idea similar to what ive been thinking
about..and make me thinkni should study more and consider to use ur engine.

Similar to what you do..
what if you etrude the 2d bodies in the 3d dimensión by the time step , x
velocity. Sort of minkowskian.  Now u have a convex hull or some spiral or
3d shape which u can tesselate as if static.  But keeping the timelike
faces only in time.  Woud it then be posible to do ccd using this extrusión
with just a static model as you say...of the time slice.  Aslo You could
rast rays and stuff on the solids 256 at a time raymarchig?  Or with ur
tree or whatever.  .like in cdd  but without 4d math, more like lighting in
3d existimg methods.  but  for real im wondering if j  should use ur engine
instead of a c# port of box2d.   With only 8 core gpu vector raycasts on a
tree and thats.slow. still  is increadibly fast in net7  i get 300 to 4000
fps woth two creatures slight wind....but  its not multiphysics and I don't
want to use 100000000 sph particles, I do oceans.  I want to do simd
raycasts or gpu on 1km space at 6cm resolution

So I'm thinking how can I use raster tech for rays on gpu w thin rigid
bodies in 2d ?
This idea came up in some 70s adn again in recent papers and they often use
6d because some geniuss   the navier stokes pressure solver in a padded
fourier space becausenits scales well.. So complex, imaginary vectors x 2 d
plus one time =6d. But the grid is 1024 x1024 not 128x128...  the results
are super fast and no upscaling tricks needed.


Now i will  getting pretyy deep, but i think u
 might intuit it.  and maybe a bit weird but I can show u the papers if u
want to be distracted .  Susskind,.Peskin, it's not quacks.Whitten,
Hawking.. all consider this model
 it's get pretty deep.into wave particle eqivalence..we gotta chose one and
as a sufer i care abour waves .  Especially long period waves.  Thry hit
like mike tyson when they break their functinn  and become partices ,the
lip.  reason is some of the physics on quantum gravity use this model  is
because e =.MC 2 not cubed, it's holograpic theory.  Mass of black hole is
also related to radius sqared.   Why? It'sAsphere?  So why not cubed?   So
was  born the hologram model..it's hard to grasp.  Even in 1 d.    No one
wants randomnes, 5d cosmology, compact dimensions,  and shrodinders CAT
doesn't.    on a more practical note,  so far, They do it for weather sims
on a 3d globe and for cosmology  to know  the geodesics so we can maybe
travel without blowijgn gas.out thr back of a crude rocket....  might sound
mad but what u are doing is what many physicists finally decided they
should...does  it make sense to you?..     and when you said "smash the
particles so that thry go in the restricted axis..which is time, youve just
done an e = mc2.    If Youve collided mass such that it is veer in to the
spatial direction only, nownits lightlike and a huge wave packet of mass
that was propagating the lattice in
In time (think tsunami scale its the speed of light ).. a tsumia was a
wavelenght.of basically the whole ocean, at 2 in high whe it goes up 90
degrees via shallpw water.massivr kdv energy, just Turned 90 degrees, into
to a bunch massive dispersive particles. A Floyd is  Mas flux.   But back
to some uranium,  it's Turned by collison, and now a
 shorter light wave and gamma and other  wave in  space converting most of
its mass. it's a flash of  light a mini sun, and possiblly Abomb. In x, y
spacelile dimensiones, only  massless  radiation can travel forever in, but
mass just fly thought time mostly until its gets near light speed it lags a
bit as in time dilation.  But collide mass  in a particel accelerator or
split and atom so it finds spacelike geodesics closer...and they exchange
form. As   it's sudenly Steer into spacelike  ,   and u either
made.some.quarks or gamma or,  an aromic bomb.  If  Mass is just  waves
packets or solitons that lag behind and travel moslty  in time dimensión,
then we don't have all the paradoxes and contractions in Gr fields.and
quantum reductivism.


  It's funny making physics games make.me read all these cosmology quantum
and and relavity papers and it's in agreement with gr but not a field
theory it's an abelian group theory.    None are verified but then again
nones got a.better gravity than gr and we know it's not explaining many
things.  Also material scientists use this stuff  for graphene which is a
similar constant, not pure but enough  to eliminate dispersion and
supercondict.


Anyways tldr but u might get approached by physicists cause even I  do and
I'm a puppet maker, no phd.  And i want to use or try their code but indont
wanna use julia ir fortran!! Or matlab. I'm trying to grt them to at try.
Net , at leasr least port c and C ++ and java  so incan use some.of their
code maybe.   Any sorry for the huge rant but I want someone to build a
flying saucer without using a supercomputing cluster thx.   I'm definite
going to take a deeper look around ir code because of the topics that come
up get more relevant. And ur voxel space thing.. sorry about the typos,
I'm so tired from coding tools.  if it doesn't make sense but u are curious
I can send u a few links to the research
